Sources/SourceKitLSP/Rename.swift

```diff
@@ -284,78 +291,438 @@ private extension RenameLocation.Usage {
+  /// `SWIFT_NAME` is used as the name translation result instead of invoking the clang importer rename rules.
+  ///
+  /// - Parameters:
+  ///   - position: A position at which this symbol is referenced from Swift.
+  ///   - snapshot: The snapshot containing the `position` that points to a usage of the clang symbol.
+  ///   - isObjectiveCSelector: Whether the name is an Objective-C selector. Cannot be inferred from the name because
+  ///     a name without `:` can also be a zero-arg Objective-C selector. For such names sourcekitd needs to know
+  ///     whether it is translating a selector to apply the correct renaming rule.
+  ///   - name: The clang symbol name.
+  /// - Returns:
+  fileprivate func translateClangNameToSwift(
+    at position: Position,
+    in snapshot: DocumentSnapshot,
+    isObjectiveCSelector: Bool,
+    name: String
+  ) async throws -> String {
+    guard let offset = snapshot.utf8Offset(of: position) else {
+      throw NameTranslationError.cannotComputeOffset(position)
+    }
+    let req = sourcekitd.dictionary([
+      keys.request: sourcekitd.requests.nameTranslation,
+      keys.sourcefile: snapshot.uri.pseudoPath,
+      keys.compilerargs: await self.buildSettings(for: snapshot.uri)?.compilerArgs as [SKDValue]?,
+      keys.offset: offset,
+      keys.namekind: sourcekitd.values.namekindObjC,
+    ])
+
+    if isObjectiveCSelector {
+      // Split the name into selector pieces, keeping the ':'.
+      let selectorPieces = name.split(separator: ":").map { String($0 + ":") }
+      req.set(keys.selectorPieces, to: sourcekitd.array(selectorPieces))
+    } else {
+      req.set(keys.baseName, to: name)
+    }
+
+    let response = try await sourcekitd.send(req, fileContents: snapshot.text)
+
+    guard let baseName: String = response[keys.baseName] else {
+      throw NameTranslationError.malformedClangToSwiftTranslateNameResponse(response)
+    }
+    let argNamesArray: SKDResponseArray? = response[keys.argNames]
+    let argNames = try argNamesArray?.map { (dict: SKDResponseDictionary) -> String in
+      guard var name: String = dict[keys.name] else {
+        throw NameTranslationError.malformedClangToSwiftTranslateNameResponse(response)
+      }
+      if name.isEmpty {
+        // Empty argument names are represented by `_` in Swift.
+        name = "_"
+      }
+      return name + ":"
+    }
+    var result = baseName
+    if let argNames, !argNames.isEmpty {
+      result += "(" + argNames.joined() + ")"
+    }
+    return result
+  }
+}
+
+/// A symbol name that can be translated between Swift and clang.
+///
+/// All properties refer to the definition symbol. For example, an Objective-C method named `performAction:with:` that
+/// is imported to Swift as `perform(action:with:)` will always have its definition in an Objective-C source file and
+/// the definition name will always be `performAction:with:`, even if rename is invoked from Swift.
+public actor TranslatableName {
```

Comment:
I do wonder if this is worth it, or if we should just do a quick iteration over the index results and pick a location for the swiftName case up front. Lazy is nice and all, but I'm not sure it matters that much here? Seems like it would avoid a lot of complexity.

Reply:
I changed the lazy name translation in `TranslatableName` to an up-front translation in https://github.com/apple/sourcekit-lsp/pull/1038/commits/0336bf43785652af5395f3f6b77e22f54c8a1a64. What do you think?